texts/0005-tauri-config-restructure.md

```diff
@@ -0,0 +1,170 @@
+- Feature Name: tauri_config_restructure
+- Start Date: 2023-13-20
+- RFC PR: [tauri-apps/rfcs#13](https://github.com/tauri-apps/rfcs/pull/13)
+- Tracking Issue: 
+
+# Summary
+
+Restructuring the `tauri.conf.json` for more simplicity, consistency and clarity.
+
+# Motivation
+
+The current `tauri.conf.jsong` has a bunch of objects that don't necessarily follow a rule that could reason why they exist there and thus makes it awkward to figure out where to add new fields
+for example `tauri > bundle` object is a configuration for the CLI, why does it exist inside `tauri` object, another example is the `plugins` object which is a tauri feature, why does it exist outside of `tauri` object? 
+These inconsistencies will make it harder to add new fields to the config. 
+
+# Guide-level explanation
+
+- Unpack `tauri` object to the root object.
+- Unpack `package` fields to the root object.
+- Move `tauri > bundle > identifier` to the root object as it is used by other places in tauri other than bundling.
+- Remove `bundle > updater` as the only useful key it has, is `pubkey` field and that should be moveed to `plugin > updater > pubkey`.
+  This will require changes in the CLI to only sign the bundles when `TAURI_SIGNING_PRIVATE_KEY` key is set .
```

Comment:
I still think reading plugin configs in the cli is fine too.
But you're right in this case the env var could work too though it could lead to "false positives" if the devs set it globally (not recommended for production of course but i do it for dev builds too since i have so many test apps 😅)

Reply:
Not really the responsibility of the CLI whether you follow best practices or not. Also the CLI shouldn't at all read the plugins configs and have specific behavior for a certain plugin without the others, this introduces a coupling between the CLI and the updater plugin and in case of breaking changes to the updater config, it will break the CLI too.

Reply:
> have specific behavior for a certain plugin without the others

It may not be the only plugin. I also talked about this a few times for the deep link plugin for example. Technically the shell plugin needs it too though it also uses an env var rn.

> in case of breaking changes to the updater config, it will break the CLI too.

Which shouldn't be a problem since we're matching the major versions.

----

The main idea here is that i believe we'll need it for more plugins than just the updater one and since we couldn't think of a general "bundler hook for plugins" thingy this is the next best approach to use as an interim solution.
Env vars likely aren't always enough and the last thing that comes to mind is to move all plugins that require bundler changes into core which a) is imho really bad (see prior discussions 😅) and b) isn't that much different from just syncing major versions implementation wise.

Reply:
So you could rephrase my initial comment as "How can we make it possible for official plugins to modify the bundler"

Reply:
It is fine for plugins to modify the CLI/Bundler behavior as long as any plugin can do so and not just selective few that we hard code into the CLI. Anyways, seems like this needs a discussion of its own so let's keep it out of this RFC.